### DISCUSSION TITLE
Zustand and Storybook

### DISCUSSION THREAD

**Discussion by sourire123** [pmndrs/zustand#2626] [Q&A] [answered]
Hi, is there any neat solution to mock Zustand store for particular story in Storybook? 

For example we can consider store with simple empty data field `name: ""` and I want to show my story with data `name: "John"`.
Thanks for help!

**Comment by dbritto-dev** [comment, marked as answer]
@sourire123 you can use storybook decorators to do that. For instance if you are using Context API you can wrap your components with a decorator to initialize the data or you can use `useEffect` and `useYourStore.setState` to do the same.

Ref: https://storybook.js.org/docs/writing-stories/decorators#context-for-mocking

**Reply by clianor** [reply]
@dbritto-dev 
Thanks for your reply.
However, it seems to me that this is very different from what I was looking for in general.
This makes it clear that the state memory is shared between stories.
Is there a way to mock the state store without sharing it between stories?

**Reply by dbritto-dev** [reply]
@clianor you can have global decorators or scoped (story/component) decorators. Ref: https://storybook.js.org/docs/writing-stories/decorators#decorator-inheritance

**Reply by clianor** [reply]
@dbritto-dev 
Unfortunately, this is not possible with global decorators or scoped decorators. The zustand store is shared across all stories.

<img width="1034" alt="image" src="https://github.com/user-attachments/assets/a964acbe-32f1-4d4b-af85-db2e1216033e" />


```ts
export const LoggedOut: Story = {
  decorators: [
    (Story) => {
      sessionStore.setState({
        session: null,
      });
      return <Story />;
    },
  ],
};

export const LoggedIn: Story = {
  decorators: [
    (Story) => {
      sessionStore.setState({
        session: {
          name: 'test',
          email: 'test@test.com',
        },
      });
      return <Story />;
    },
  ],
};
```

<img width="1022" alt="image" src="https://github.com/user-attachments/assets/8c21115f-6342-478b-a6cd-7823222f8259" />

```ts
export const LoggedOut: Story = {};

export const LoggedIn: Story = {
  decorators: [
    (Story) => {
      sessionStore.setState({
        session: {
          name: 'test',
          email: 'test@test.com',
        },
      });
      return <Story />;
    },
  ],
};
```

**Reply by clianor** [reply]
Although it wasn't what I was looking for, thank you for your quick response.

**Reply by dbritto-dev** [reply]
@clianor about store is shared across all stories is expected due to you are using global stores on each story/component

**Reply by dbritto-dev** [reply]
keep in mind that you can create scoped (context based) stores in zustand too

**Reply by clianor** [reply]
@dbritto-dev 
thank you I didn't want to use that method, but it seemed like there was no other way.

**Reply by dbritto-dev** [reply]
btw, I wouldn't use the same store in your case I'd create stores (even the global stores) for each story/component if is necessary

**Reply by clianor** [reply]
@dbritto-dev Thanks for the good suggestion.
